Update "Running large tests" article (#992)
grafana/k6-docs: src/data/markdown/translated-guides/en/06 Testing Guides/04 Running large tests.md

excerpt: 'How to run large-scale k6 tests without distributed-execution'
---

> A common misconception about load testing is that it requires [distributed execution](#distributed-execution) (that is, load generated from multiple machines).
> This isn't the case with k6.

Reviewer: This is a very strong statement that arguably is wrong. 

Can we have not so marketing message of 
```suggestion
A common misconception about load testing is that it requires [distributed execution](#distributed-execution) (that is, load generated from multiple machines).
This isn't necessarily the case with k6.
```

Reviewer 2: I support changing the wording here and in other places. 

Haven't reviewed it. @mstoykov, it can help if you can point other sentences.

Author: Hhmm what would you say is wrong about it? That it's not a common misconception, or that k6 does require distributed execution?

I'm not sure how common of a misconception it is, but k6 certainly doesn't _require_ it. I'd say there's some truth in both statements.

I can add necessarily to soften this, but maybe we should rephrase this altogether? WDYT @MattDodsonEnglish?

Reviewer 3: How about this?

> k6 can generate a lot of load from a small machine.
> With proper monitoring and script optimization, you might be able to run a rather large load test without needing distributed execution.

Author: Thanks, Matt!

I changed this, and a few other places that were too harsh.

@mstoykov Let us know if 62bd7e8d is an improvement.

Reviewer: For me that changes is great. 

IMO it was wrong on a lot of levels but especially on the hardness of "it isn't the case".

But distributed mode isn't just used for *scale* it also used to make the test come from multiple (usually) geo distributed machines. 

And we are not doing some very complex test - for something more complex we will need even more resources.

Sometimes there are other considerations (as the article already goes through) such as the number of ports and the need for multiple IPs. 

Arguably those are fixable *without* distributed execution, but probably only to an extend.

So the whole hardness of the "it isn't the case" is the *biggest* problem.

> This document explains how to launch a large-scale k6 test on a single machine without the need for distributed execution.

Reviewer 3: Maybe we should prepare readers for what they are going to read and what they need?

```
This document explains how to launch a large-scale k6 test on a single machine without the need for distributed execution. It most cases, the process involves the following:
- Shell commands to increase the default network and user limits of the operating system
- Monitoring to ensure safe resource usage from the load generator
- Efficient test design, with attention to scripting, k6 options, and file uploads.
- Some error monitoring to detect network and SUT failures.
```

Something like this would serve as a TL;DR, and it would prepare readers for what they need to do (as it's more than just adding a flag,  `--iterations 1000000`)

Author: Sure, see 8d725942. I changed some of the wording slightly.

A common misconception of many load testers is that [distributed execution](#distributed-execution) (the ability to launch a load test from multiple machines) is required to generate large load. This is not the case with k6.

k6 is different from many other load testing tools in the way it handles hardware resources. A single k6 process will efficiently use all CPU cores on a load generator machine. Depending on the available resources, and taking into consideration some of the advice below, a single instance of k6 is able to run 30,000-40,000 simultaneous users (VUs). In some cases this number of VUs can generate up to 300,000 HTTP requests per second (RPS).

Unless you need more than 100,000-300,000 requests per second (6-12M requests per minute), a single instance of k6 will likely be sufficient for your needs.

Below we will explore what hardware and considerations are needed for generating different levels of load.

> ## OS fine-tuning

Reviewer 3: As a casual, I feel uncomfortable running these commands.

These processes require on root and it looks like they modify kernal interaction and pre-defined user resource limits. As a reader, I think "why are these limits here? Am I doing something dangerous?" As a writer, I think, "Am I encouraging readers to copy and paste something without thinking about potentially serious ramifications?"

I looked at the "Fine Tuning OS" and see there's a bit more guidance. Still, though this section links that page, I'd suggest putting an admonition in this section.

```
<Blockquote mod="attention" title="">

The default limits protect your system from programs that might leak memory in huge quantities.
Consider increasing the limits only for the shell session that runs the test.

</Blockquote>

```

Maybe the language isn't correct, but something like that. Also, I see that the `ulimit` command by default doesn't persist, does the `sysctl` configuration change persist? If so, it might be good to provide details for reverting it.

Author: > As a casual, I feel uncomfortable running these commands.

Well, this article is directed more towards advanced k6 users, who are reaching the limits imposed by their OS. I agree that it should be as friendly as the rest of our documentation, but at some point we have to expect some familiarity from the reader.

We do explain the purpose of these commands below, and that they require root privileges (only the `sysctl` commands do, but we gloss over some of these subtleties that are better explained in the "Fine tuning OS" article). These are not "dangerous" by any means, but tweak the system to allow k6 (and any other program) to achieve maximum performance. These settings differ from OS and Linux distribution, but are usually set rather low by default, since the general public doesn't typically run programs that use tens of GBs of RAM, and opens thousands of concurrent TCP connections.

Also note that the settings changed by these commands are temporary, and will be reset to the system default after a reboot. If the user wants to make them permanent (presumably after they've confirmed the improvement), they can change some system files, which also require root privileges. The `ulimit` command applies only for the current shell session, so it has even less impact. This is all better explained in the "Fine tuning OS" article.

So I'm not sure what kind of warning would be helpful here, given the audience this article is aimed towards.

Reviewer 3: > Well, this article is directed more towards advanced k6 users, 

Two things here: the article doesn't state directly who it's for (but the new intro does a much better job of setting expectations :+1:); how much does "advanced k6" use imply Linux knowledge?

I don't see what information with these extra two sentences.

```
These default limits are in place because ordinary programs don't need to open thousands of concurrent TCP connections or use dozens of GBs of RAM.
The `ulimit` changes apply only for the current shell, and the other settings reset when the system reboots.
For details, refer to [Fine-tuning OS](/misc/fine-tuning-os). 
```

Author: > the article doesn't state directly who it's for

None of our articles mention the intended audience AFAIK, but whoever is interested in "running large tests" and reaching the limits of their OS, would likely expect to read about advanced topics.

> how much does "advanced k6" use imply Linux knowledge?

I initially wrote "advanced user" in general, and still think it applies. I think there's some overlap between an advanced k6 user and an advanced computer user, so we don't need to explain every topic from scratch, especially since we already mention these details on the "Fine tuning OS" page.

That said, I expanded the intro a bit in 7e64d232. Let me know if this is an improvement.

> The following OS changes for GNU/Linux allow k6 to use the **full network capacity** of the machine for maximum performance.

Reviewer 3: ```suggestion
The following OS changes for GNU/Linux let k6 use the **full network capacity** of the machine for maximum performance.
```

Regardless of the test file, you can assume that large tests require a significant amount of CPU power.
We recommend that you size the machine to have at least 20% idle cycles (up to 80% used by k6, 20% idle).
If k6 uses 100% of the CPU to generate load, the test will experience throttling limits, and this may
cause the result metrics to have a much larger response time than in reality.

Memory
: k6 likes memory, but isn't [as greedy as other load testing tools](https://k6.io/blog/comparing-best-open-source-load-testing-tools#memory-usage).
Memory consumption heavily depends on your test scenarios. To estimate the memory requirement of your test,
run the test on your development machine with 100VUs and multiply the consumed memory by the target number of VUs.
: Simple tests use ~1-5MB per VU. (1000VUs = 1-5GB).
Tests that use file uploads, or load large JS modules, can consume tens of megabytes per VU.
Note that each VU has a copy of all JS modules your test uses.
To share memory between VUs, consider using [SharedArray](/javascript-api/k6-data/sharedarray/), or an external data store, such as [Redis](/javascript-api/k6-experimental/redis/).

> : If you're using [swap space](https://en.wikipedia.org/wiki/Memory_paging), consider disabling it. If the system runs out of physical memory, the process of swapping memory to much slower secondary storage will have erratic effects on performance and system stability. Instead, plan ahead for the memory usage you expect your tests to achieve, and ensure that you have enough physical RAM for the usage to not exceed 90%.

Reviewer: ```suggestion
: If you're using [swap space](https://en.wikipedia.org/wiki/Memory_paging), consider disabling it. If the system runs out of physical memory, the process of swapping memory to much slower secondary storage will have erratic effects on performance and system stability. Which likely will invalidate any test results as the load generator had different performance in different parts of the test. Instead, plan ahead for the memory usage you expect your tests to achieve, and ensure that you have enough physical RAM for the usage to not exceed 90%.
```

It is probably good idea to spell out what the consequence for the testing is

run the test on your development machine with 100VUs and multiply the consumed memory by the target number of VUs.

Simple tests will use ~1-5MB per VU. (1000VUs = 1-5GB).
Tests that are using file uploads, or load large JS modules, can consume tens of megabytes per VU.
Keep in mind that each VU has a copy of all JS modules your test uses.
If you need to share memory between VUs, consider using [SharedArray](/javascript-api/k6-data/sharedarray/), or an external data store, such as [Redis](/javascript-api/k6-experimental/redis/).

## General advice

> ### Make your test code resilient

Reviewer 3: ```suggestion
### Resilient error handling
```

I think we should give headings names that are as precise as possible.

When running large stress tests, your script can't assume anything about the HTTP response.

> Often performance tests are written with a "happy path" in mind.

Reviewer 3: ```suggestion
A large test might elicit unexpected behavior from the SUT, so your test script should handle exceptions well.

In testing literature, the correct, predictable response is often called the [_Happy path_](https://en.wikipedia.org/wiki/Happy_path).
A common mistake in performance testing happens when the script assumes that SUT will take the happy path.
```

Reviewer 3: I'd seen this "Happy path" term twice now, and I only just realized that it's an industry term. I think we should define it. Now I'm not the brightest, but even if it's a common term in CS, not all of our readers will come from the same language and educational backgrounds.

Author: It's a pretty common term in testing, even outside of load/performance testing.

Instead of defining it here, let's add it to the [glossary](https://k6.io/docs/misc/glossary/)? Then we can link it here, just as we did with SUT.

> For example, a "happy path" check like the one below is something that we often see in k6 scripts.

Reviewer 3: Now that I understand what "Happy path" is, is this really a happy path check? It seems like it does not assume the happy path, since it accepts the possibility of an incorrect body size. 

Instead, it seems like it's more a case of error handling that's not general enough.

Proposed change:

```
Even if scripts accept the SUT won't always respond according to the happy path, it still might not handle a sufficiently wide range of errors. 
For example, the following check is something we often see in k6 scripts:

Author: It is a happy path check, since it doesn't take into account failure conditions under load, when the response body wouldn't exist. The body size check is a standard assertion, and the user might want the test to fail if it's incorrect.

Your suggestion sounds more confusing, sorry, so I'd rather leave this as is.

Reviewer 3: > It is a happy path check, since it doesn't take into account failure conditions under load, when the response body wouldn't exist. The body size check is a standard assertion, and the user might want the test to fail if it's incorrect.

I don't understand. Isn't an incorrect body size another indication of system failure? Or are we defining failure as "no body response at all?". From the Wikipedia page, I thought a happy path is where there are no error conditions.

});
```

Code like this runs fine when the [system under test](/misc/glossary/#system-under-test) (SUT) is not overloaded and returns proper responses.
When the system starts to fail, the above check won't work as expected.

The issue here is that the check assumes that there's always a body in a response. The `r.body` may not exist if the server is failing.

> In such case, the check itself won't work as expected and an error similar to the one below will be returned:

Reviewer 3: Shouldn't we add an example of how to handle this exception? Right now it seems we diagnose a problem, but don't give any way to find a solution.

Reviewer 3: Oh, I realize now that this is in the text, just not in the diffs. Still, it looks like the example only fails a check if there is a body to begin with.

```javascript
const checkRes = check(res, {
  'Homepage body size is 11026 bytes': (r) => r.body && r.body.length === 11026,
});
```

Wouldn't it make more sense to have two checks? One that checks that a body is returned and other that, when returned, it's the correct length?

Author: Matt, I think you're nitpicking at this point. It's fine as is :)

</CodeGroup>

### Monitor the load generator machine

If you are running a test for the first time, it's a good idea to keep an eye on the available resources while the test is running.
The easiest way to do so is to start at least 2 terminals one the machine: one to run k6, and others to monitor the CPU, memory and network usage.

We recommend the following command-line tools for monitoring resources:

> - CPU and memory: [htop](https://en.wikipedia.org/wiki/Htop) or [nmon](https://nmon.sourceforge.net/).
> 
> - Network: [iftop](https://en.wikipedia.org/wiki/Iftop) or [nmon](https://nmon.sourceforge.net/).

Reviewer 3: ```suggestion
- CPU and memory: [`htop`](https://en.wikipedia.org/wiki/Htop) or [`nmon`](https://nmon.sourceforge.net/).

- Network: [`iftop`](https://en.wikipedia.org/wiki/Iftop) or [`nmon`](https://nmon.sourceforge.net/).
```

Reviewer 3: Try to put code and terminal programs in `monospace`, which gives the reader a little clue about the essential form of whatever's being described.

If you prefer graphical applications, you can use your OS's system monitoring tool (e.g. GNOME System

> Monitor or Plasma System Monitor), or use a standalone tool like [SysMonTask](https://github.com/KrispyCamel4u/SysMonTask).

Reviewer 3: Maybe we could put a little summary repetition of what the reader should be looking for here?

Something like:

```
As general guidelines, look for the the following:
- CPU utilization doesn't exceed 80%
- Network traffic isn't constant (?)
- Memory doesn't approach full utiliazation (?)

Reviewer 3: I put (?) when I'm not sure.

Author: Let me know if 63f1f8e9 resolves this.

We recommend the following command-line tools for monitoring resources:

- CPU and memory: [htop](https://en.wikipedia.org/wiki/Htop) or [nmon](https://nmon.sourceforge.net/).

- Network: [iftop](https://en.wikipedia.org/wiki/Iftop) or [nmon](https://nmon.sourceforge.net/).

If you prefer graphical applications, you can use your OS's system monitoring tool (e.g. GNOME System
Monitor or Plasma System Monitor), or use a standalone tool like [SysMonTask](https://github.com/KrispyCamel4u/SysMonTask).

Here's a screenshot of 3 terminal sessions showing k6, iftop and htop.
![k6 iftop htop](./images/large-scale-testing-3-terminals.png)

As general guidelines, look for the following:
- CPU utilization doesn't exceed 90%. If all CPU cores are 100% utilized during the test run, you might notice performance degradation in your test results.
- Network utilization is at an acceptable level. Depending on your test, you might expect to fully saturate the network bandwidth, or this might be a signal that your test is bound by the available bandwidth. In other scenarios, you might want to minimize network usage to keep costs down.

> - Memory utilization doesn't exceed 90%. If you're close to exhausting available physical RAM, the system might start swapping memory to disk, which will affect performance and system stability. In extreme cases, running out of memory on Linux will cause the system to end the k6 process.

Author: Should we recommend running without swap? Our loadgens don't use it, AFAIK. cc: @mstoykov

Reviewer: I would recommend people to run without swap *in general* ;) The only use at this point IMO are : 
1. systems which don't have enough ram at all
2. hibernation (which is why I still have swap on my laptop)

In (my) practice the difference between rogue process running out of memory and dying and going into swap is just that in the second case it dies a lot slower and later usually thrashing the whole system in the process ;). 

I guess with NVMe SSDs this is a bit better but even on my laptop whenever I go to swap I find out as the whole system goes really slow.

Given that k6 is a performance tool - it performance degrading significantly because swap is used seems like a terrible idea even if the process itself doesn't die.

Author: Fully agree, so I think this deserves a separate section that explains it. Will add, thanks.

Author: Resolved by 1ed89f2c.

The k6 settings listed below will unlock additional performance benefits when running large tests.

> ### discardResponseBodies

Reviewer 3: I need to think about whether to put these in `monospace` . I'm leaning towards yes, even though it's unsightly.

Author: I don't think there's an objective best practice here, regardless of what some authority says. :smile:

Consistency matters more, so if you strongly prefer it this way, feel free to change it, and I'll make sure to use it from now on.

k6 by default loads the response body of the request into memory. This causes much higher memory consumption and is often not necessary.
You can tell k6 to not process the body of all responses by setting `discardResponseBodies` in the options object like this:

```javascript
export const options = {
  discardResponseBodies: true,
};
```

If you need the response body for some requests you can override this with [Params.responseType](/javascript-api/k6-http/params).

### --no-thresholds --no-summary

> If you are running a local test and streaming results to the k6 Cloud (`k6 run -o cloud`), you may want to disable the terminal summary
> and local threshold calculation because thresholds and summary will be displayed in the Cloud.

Reviewer 3: ```suggestion
If you are running a local test and streaming results to the k6 Cloud (`k6 run -o cloud`), you may want to disable the terminal summary
and local threshold calculation,
because k6 Cloud will display the summary and calculate the thresholds.

Without these options, the operations will be duplicated by both the local machine and the k6 Cloud servers.
```

This will save you some memory and CPU cycles.

Here are all the mentioned flags, all in one:

```bash
k6 run scripts/website.js \
  -o cloud \
  --vus=20000 \
  --duration=10m \
  --no-thresholds \
  --no-summary
```

> ### Test script optimizations

Reviewer 3: Shouldn't this be its own H2? It seems incoherent to be nested under "Options"

> If you are trying to squeeze more performance out of the hardware, you can consider optimizing the code of the test script itself.

Reviewer 3: I think each of these sections should mention how the proposed change would manifest in terms of system performance. If all improve memory, then we could retitle the heading: "Script optimizations to reduce memory"

Reviewer 3: Also, some of these features are quite important. It raises the question, if users have to disable them, is it worth running the test. Why should someone run such a test?

#### Checks and groups

k6 records the result of every individual check and group separately. If you are using many checks and groups, you may consider removing them to boost performance.

#### Custom metrics

Similar to checks, values for custom metrics (Trend, Counter, Gauge and Rate) are recorded separately. Consider minimizing the usage of custom metrics.

> #### Thresholds with abortOnFail

Reviewer 3: Is this a CPU cost?

If you have configured [abortOnFail thresholds](/using-k6/thresholds#aborting-a-test-when-a-threshold-is-crossed), k6 needs to evaluate the result constantly to verify that the threshold wasn't crossed. Consider removing this setting.

#### URL grouping

k6 v0.41.0 introduced a change to support metric time-series. A side-effect of this is that every unique URL will create a new time-series object, which may consume more RAM than expected.

The solution to this is to use the [URL grouping](https://k6.io/docs/using-k6/http-requests/#url-grouping) feature.

#### JavaScript optimizations

Finally, if all of the above suggestions are insufficient, there might be some JavaScript optimizations you can do. This includes general improvements to minimize script complexity: avoid deeply nested `for` loops, don't keep references to large objects in memory if it can be avoided, keep external JS dependencies to a minimum, perform tree shaking of the k6 script if you have a build pipeline, etc.

> See [this article](https://javascript.info/garbage-collection) about garbage collection in the V8 runtime. While the JavaScript VM k6 uses is very different and runs on Go, the general principles are applicable. Keep in mind that memory leaks are still possible in k6 scripts, and might lead to much quicker RAM exhaustion if not fixed.

Reviewer 3: ```suggestion
Refer to this article about [garbage collection](https://javascript.info/garbage-collection) in the V8 runtime. While the JavaScript VM k6 uses is very different and runs on Go, the general principles are applicable. Keep in mind that memory leaks are still possible in k6 scripts, and might lead to much quicker RAM exhaustion if not fixed.
```

k6 needs a significant amount of memory when uploading files, as every VU is independent and has its own memory.

### Costs

> #### Data transfer

Reviewer 3: Getting into `h4s` is kind of "prose smell". So much nesting makes it hard for the reader to understand what this relates to.

In any case, we can at least use description lists: https://github.com/grafana/k6-docs/blob/main/CONTRIBUTING/gatsby-reference.md#description-list

Elsewhere, it probably makes sense to flatten the structure. I'll look into this.

Reviewer 3: > saw some examples of how description lists are rendered, and I'm not sure if they would be a good fit here. > The titles are too large, and we don't need the indentation of the bodies.

I agree, I wont to reduce the size of description list items, but that's a CSS change. Anyway, they're pretty much the size of the h4s so no difference there.

The indentation is fine, IMO: it creates a clear sequence of repetition and alignment. With h4 there's only one level of alignment, the definition makes a clearer sequence of Item and description.

It would look like this: https://k6.io/blog/reverse-load-test/#analysis-bottlenecks-and-planned-improvements

Really I think it's fine.

Anyway the docs table of contents shows only h2s, so the best way to improve that navigation is by flattening the structure out, reducing h3s and h4s.

```bash
WARN[0064] Request Failed    error="Get http://test.k6.io: context deadline exceeded"
```

### dial tcp 52.18.24.222:80: i/o timeout

This is a similar error to the one above, but in this case, k6 wasn't even able to make an HTTP request. The target system isn't able to establish a TCP connection.

```bash
WARN[0057] Request Failed     error="Get http://test.k6.io/: dial tcp 52.18.24.222:80: i/o timeout"
```

### socket: too many open files

This error means that the load-generator machine isn't able to open TCP sockets because it reached the limit of open file descriptors.
Make sure that your limit is set sufficiently high. See our ["Fine tuning OS" article](/misc/fine-tuning-os/#viewing-limits-configuration).

```bash
WARN[0034] Request Failed     error="Get http://test.k6.io/: dial tcp 99.81.83.131:80: socket: too many open files"
```

Note: you should decide what level of errors is acceptable. At large scale, some errors are always present.
If you make 50M requests with 100 failures, this is generally a good result (0.00002% errors).

## Benchmarking k6

> We have executed a few large tests on different EC2 machines to see how much load k6 can generate. Our general observation is that k6 scales proportionally to the hardware. A machine with 2x more resources can generate roughly 2x more traffic. The limit to this scalability is in the number of open connections. A single Linux machine can open up to 65,535 sockets per IP. This means that a maximum of 65k requests can be sent simultaneously on a single machine. The RPS limit depends on the response time of the SUT. If responses are delivered in 100ms, the RPS limit is 650,000.

Reviewer 3: > A single Linux machine can open up to 65,535 sockets per IP. 

Could we site a source here?

Reviewer 3: I agree with RFC for authority.

> We maintain a [repository](https://github.com/grafana/k6-hardware-benchmark) with some scripts used to benchmark k6 and create reports. These tests are run for every new k6 version, and you can see the results in the [`results/` directory](https://github.com/grafana/k6-hardware-benchmark/tree/master/results).

Author: Nice catch, but like I mentioned in the PR description, we shouldn't merge this before the benchmarking is finished. That will happen after the results are added to the k6-hardware-benchmark repo, so this link will eventually work.
